nexus/src/app/disk.rs

```diff
@@ -408,4 +352,277 @@ impl super::Nexus {
+            //
+            // In this scenario the write here would fail because the volume
+            // would have been detached from the Pantry, but say that the user
+            // instead called bulk-write-stop, then import, thereby kicking off
+            // the import blocks from URL saga? The Pantry locks its internal
+            // entry by the ID of the volume, so at least the requests would be
+            // ordered by their arrival time: either this bulk write would land
+            // first, or the import would land first:
+            //
+            // - if this bulk write landed first, the import would likely
+            //   overwrite it with blocks from a URL.
+            //
+            // - if the import landed first, then the bulk write would overwrite
+            //   what the import job imported, probably corrupting the data.
+            //
+            // I also believe it's not correct to use a saga here. Really, the
+            // user's cli (or any other program really) is responsible for this
+            // bulk write operation, not Nexus. If the bulk_write call below
+            // fails, then that failure would be propagated up to the user, and
+            // that user's program can act accordingly. In a way, the user's
+            // program is an externally driven saga instead.
+
+            let client = crucible_pantry_client::Client::new(&format!(
+                "http://{}",
+                endpoint
+            ));
+            let request = crucible_pantry_client::types::BulkWriteRequest {
+                offset: param.offset,
+                base64_encoded_data: param.base64_encoded_data,
+            };
+
+            client
+                .bulk_write(&db_disk.id().to_string(), &request)
+                .await
+                .map_err(|e| match e {
+                    crucible_pantry_client::Error::ErrorResponse(rv) => {
+                        match rv.status() {
+                            status if status.is_client_error() => {
+                                Error::invalid_request(&rv.message)
+                            }
+
+                            _ => Error::internal_error(&rv.message),
+                        }
+                    }
+
+                    _ => Error::internal_error(&format!(
+                        "error sending bulk write to pantry: {}",
+                        e,
+                    )),
+                })?;
+
+            Ok(())
+        } else {
+            error!(self.log, "disk {} has no pantry address!", db_disk.id());
+            Err(Error::internal_error(&format!(
+                "disk {} has no pantry address!",
+                db_disk.id(),
+            )))
+        }
+    }
+
+    /// Move a disk from the "ImportingFromBulkWrites" state to the
+    /// "ImportReady" state, usually signalling the end of manually importing
+    /// blocks.
+    pub async fn disk_manual_import_stop(
+        self: &Arc<Self>,
```

Comment:
Right - this slightly protects the user from a race where one thread is still bulk writing in, and another calls finalize.

Imagine two threads, one performing bulk writes, and another erroneously calling finalize. If it's possible to transition from `ImportingFromBulkWrites` to `Finalizing`, then thread 1 will error, but thread 2 will succeed: it's interrupting what thread 1 is doing but that state transition is allowed. The disk will be successfully finalized but will fail the checksum validation (only if one is supplied!).

With an explicit transition back to `ImportReady` required, thread 2 will fail to finalize the disk unless it calls `bulk-write-stop` first.

Note if thread 1 is performing bulk writes and thread 2 erroneously calls `bulk-write-stop` then `finalize`, it'll still interrupting what thread 1 is doing and will successfully finalize (still failing the checksum validation if it's supplied).

In the case of importing from a url, a saga is changing the disk's state. It won't be possible to finalize the disk until the pantry job is complete because the saga will set the disk's state back to import ready when the job is done. If we could transition from `ImportingFromUrl` to `Finalizing`, then users could cut the saga off and/or cause the saga to unwind. In this case, the fact that saga transitioned from `ImportReady` to `ImportingFromUrl` means that it's "locked" the disk and has exclusive use of it because all operations start only from state `ImportReady`, including finalizing.